python/lsst/ts/ofc/controllers/base_controller.py

```diff
@@ -0,0 +1,305 @@
+    freedom (DOF) at time `k+1` based on the wavefront error (`yk`) at time
+    `k`.
+
+    Parameters
+    ----------
+    ofc_data : `OFCData`
+        Data container.
+    log : `logging.Logger` or `None`
+        Optional logging class to be used for logging operations. If `None`,
+        creates a new logger.
+
+    Attributes
+    ----------
+    dof_state : `np.array`
+        State of telescope in the basis of degrees of freedom.
+    dof_state0 : `np.array`
+        Initial state of telescope in the basis of degrees of freedom.
+    log : `logging.Logger`
+        Logger class used for logging operations.
+    ofc_data : `OFCData`
+        OFC data container.
+    """
+
+    # Eta in FWHM calculation.
+    ETA = 1.086
+
+    # FWHM in atmosphere.
+    FWHM_ATM = 0.6
+
+    def __init__(self, ofc_data: OFCData, log: logging.Logger | None = None) -> None:
+        if log is None:
+            self.log = logging.getLogger(type(self).__name__)
+        else:
+            self.log = log.getChild(type(self).__name__)
+
+        # Set OFC data
+        self.ofc_data = ofc_data
+
+        # Set gains and setpoint
+        self._kp = self.ofc_data.controller["kp"]
+        self._ki = self.ofc_data.controller["ki"]
+        self._kd = self.ofc_data.controller["kd"]
+        self.setpoint = np.array(self.ofc_data.controller["setpoint"])
+
+        # Set initial state
+        self.dof_state0 = np.zeros(len(self.ofc_data.dof_idx))
+
+        for comp in self.ofc_data.comp_dof_idx:
+            start_idx = self.ofc_data.comp_dof_idx[comp]["startIdx"]
+            end_idx = (
+                self.ofc_data.comp_dof_idx[comp]["startIdx"]
+                + self.ofc_data.comp_dof_idx[comp]["idxLength"]
+            )
+
+            state0name = self.ofc_data.comp_dof_idx[comp]["state0name"]
+
+            if "Hexapod" in state0name:
+                self.dof_state0[start_idx:end_idx] = np.array(
+                    [
+                        self.ofc_data.dof_state0[state0name][axis_name]
+                        for axis_name in ["dZ", "dX", "dY", "rX", "rY"]
+                    ]
+                )
+            else:
+                self.dof_state0[start_idx:end_idx] = np.array(
+                    [
+                        self.ofc_data.dof_state0[state0name][f"mode{mode_idx+1}"]
+                        for mode_idx in range(
+                            self.ofc_data.comp_dof_idx[comp]["idxLength"]
+                        )
+                    ]
+                )
+
+        self.dof_state = self.dof_state0.copy()
+
+        # Initialize previous error and integral
+        self.previous_error = self.setpoint - self.dof_state0
+        self.integral = self.previous_error
+
+        # Initialize PSSN data
+        self.pssn_data = dict(sensor_names=None, pssn=None)
+
+    @property
+    def kp(self):
+        return self._kp
+
+    @kp.setter
+    def kp(self, value):
+        if 0.0 <= value <= 1.0:
+            self._kp = value
```

Comment:
I have removed this constraint and the corresponding unit test function that tested it.

Reply:
Yep. The following link should explain why we could remove the restriction:

[File:PID Compensation Animated.gif](https://en.wikipedia.org/wiki/File:PID_Compensation_Animated.gif)

And since we might have the overshoot in the tuning of PID parameters, it is common to add the saturator. For example, you could add the restriction for the final output `uk` should be less than the `error` or 2 * `error` or others (you can put the parameter to avoid to hard-code anything).

Thanks!